backend/test/core_service_impl_test.cpp

```diff
@@ -0,0 +1,278 @@
+        EXPECT_EQ(response.plugin_info(i).address(), DEFAULT_BACKEND_ADDRESS);
+    }
+}
+
+TEST_F(CoreServiceImplTest, portIsDefaultInPluginInfos)
+{
+    dronecore::rpc::core::ListRunningPluginsResponse response;
+
+    _core_service->ListRunningPlugins(nullptr, nullptr, &response);
+
+    for (int i = 0; i < response.plugin_info_size(); i++) {
+        EXPECT_EQ(response.plugin_info(i).port(), DEFAULT_BACKEND_PORT);
+    }
+}
+
+TEST_F(CoreServiceImplTest, subscribeDiscoverActuallySubscribes)
+{
+    EXPECT_CALL(*_dc, register_on_discover(_))
+    .Times(1);
+    grpc::ClientContext context;
+    dronecore::rpc::core::SubscribeDiscoverRequest request;
+
+    _stub->SubscribeDiscover(&context, request);
+    _core_service->stop();
+}
+
+TEST_F(CoreServiceImplTest, discoverStreamEmptyIfCallbackNotCalled)
+{
+    std::vector<uint64_t> uuids;
+    auto uuids_stream_future = subscribeDiscoverAsync(uuids);
+
+    _core_service->stop();
+    uuids_stream_future.wait();
+
+    EXPECT_EQ(0, uuids.size());
+}
+
+std::future<void> CoreServiceImplTest::subscribeDiscoverAsync(std::vector<uint64_t> &uuids)
+{
+    return std::async(std::launch::async, [&]() {
+        grpc::ClientContext context;
+        dronecore::rpc::core::SubscribeDiscoverRequest request;
+        auto response_reader = _stub->SubscribeDiscover(&context, request);
+
+        dronecore::rpc::core::DiscoverResponse response;
+        while (response_reader->Read(&response)) {
+            uuids.push_back(response.uuid());
+        }
+    });
+}
+
+TEST_F(CoreServiceImplTest, discoverSendsOneUUID)
+{
+    const int expected_uuid = 42;
+    std::promise<void> subscription_promise;
+    auto subscription_future = subscription_promise.get_future();
+    dronecore::testing::event_callback_t event_callback;
+    EXPECT_CALL(*_dc, register_on_discover(_))
+    .WillOnce(SaveCallback(&event_callback, &subscription_promise));
+
+    std::vector<uint64_t> uuids;
+    auto uuids_stream_future = subscribeDiscoverAsync(uuids);
+    subscription_future.wait();
+    event_callback(expected_uuid);
+    _core_service->stop();
+    uuids_stream_future.wait();
+
+    EXPECT_EQ(1, uuids.size());
+    EXPECT_EQ(expected_uuid, uuids.at(0));
+}
+
+TEST_F(CoreServiceImplTest, discoverSendsMultipleUUIDs)
+{
+    const int uuid0 = 234132413;
+    const int uuid1 = 948789299;
+    const int uuid2 = 861987343;
+    std::promise<void> subscription_promise;
+    auto subscription_future = subscription_promise.get_future();
+    dronecore::testing::event_callback_t event_callback;
+    EXPECT_CALL(*_dc, register_on_discover(_))
+    .WillOnce(SaveCallback(&event_callback, &subscription_promise));
+
+    std::vector<uint64_t> uuids;
+    auto uuids_stream_future = subscribeDiscoverAsync(uuids);
+    subscription_future.wait();
+    event_callback(uuid0);
+    event_callback(uuid1);
+    event_callback(uuid2);
+    _core_service->stop();
+    uuids_stream_future.wait();
+
+    EXPECT_EQ(3, uuids.size());
```

Comment:
Nitpick, this would have to be an `ASSERT_EQ` because the `at()` below will fail otherwise.

Reply:
Good point :)